File: java/src/main/java/com/authdog/resources/PortalResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * User portal operations.
 */
public final class PortalResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a portal helper.
     * @param clientParam Authdog client
     */
    public PortalResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * Generate a portal magic link.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return link envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode generateLink(final String tenantId,
                                 final String environmentId,
                                 final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/portal/generate-link",
                body, null, JsonNode.class);
    }
}

File: csharp/Resources/PortalResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// User-portal Wave 3 operations
    /// </summary>
    public class PortalResource
    {
        private readonly AuthdogClient _client;

        public PortalResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> GenerateLinkAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/portal/generate-link",
                body);

        public JObject GenerateLink(string tenantId, string environmentId, object body) =>
            GenerateLinkAsync(tenantId, environmentId, body).GetAwaiter().GetResult();
    }
}
